x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authc/RealmsTests.java

```diff
@@ -162,6 +160,259 @@ private void setRealmAvailability(Function<LicensedFeature.Persistent, Boolean>
         licenseStateListeners.forEach(LicenseStateListener::licenseStateChanged);
     }
 
+    public void testReservedRealmCannotBePartOfDomain() {
+        Settings.Builder builder = Settings.builder()
+            .put("path.home", createTempDir())
+            .put(Node.NODE_NAME_SETTING.getKey(), randomAlphaOfLengthBetween(3, 8));
+        String domainName = randomAlphaOfLength(7);
+        String nativeRealmName = randomFrom("n" + randomAlphaOfLength(8), NativeRealmSettings.DEFAULT_NAME);
+        builder.put("xpack.security.authc.realms.native." + nativeRealmName + ".enabled", randomBoolean());
+        builder.put("xpack.security.authc.realms.native." + nativeRealmName + ".order", 4);
+        String fileRealmName = randomFrom("f" + randomAlphaOfLength(8), FileRealmSettings.DEFAULT_NAME);
+        builder.put("xpack.security.authc.realms.file." + fileRealmName + ".enabled", randomBoolean());
+        builder.put("xpack.security.authc.realms.file." + fileRealmName + ".order", 5);
+        builder.put(
+            "xpack.security.authc.domains." + domainName + ".realms",
+            randomFrom(nativeRealmName + "," + ReservedRealm.NAME, ReservedRealm.NAME + "," + fileRealmName)
+        );
+        Settings settings = builder.build();
+        Environment env = TestEnvironment.newEnvironment(settings);
+        IllegalArgumentException e = expectThrows(
+            IllegalArgumentException.class,
+            () -> new Realms(settings, env, factories, licenseState, threadContext, reservedRealm)
+        );
+        assertThat(e.getMessage(), containsString("Undefined realms [reserved] cannot be assigned to domains"));
+    }
+
+    public void testReservedRealmRef() throws Exception {
+        Settings.Builder builder = Settings.builder()
+            .put("path.home", createTempDir())
+            .put(Node.NODE_NAME_SETTING.getKey(), randomAlphaOfLengthBetween(3, 8));
+        String nativeRealmName = randomFrom("n" + randomAlphaOfLength(8), NativeRealmSettings.DEFAULT_NAME);
+        builder.put("xpack.security.authc.realms.native." + nativeRealmName + ".enabled", randomBoolean());
+        builder.put("xpack.security.authc.realms.native." + nativeRealmName + ".order", 4);
+        String fileRealmName = randomFrom("f" + randomAlphaOfLength(8), FileRealmSettings.DEFAULT_NAME);
+        builder.put("xpack.security.authc.realms.file." + fileRealmName + ".enabled", randomBoolean());
+        builder.put("xpack.security.authc.realms.file." + fileRealmName + ".order", 5);
```

Comment:
Optional since that's really not the focus of the PR, but for coverage we might also randomly disable the reserved realm via `builder.put("xpack.security.authc.reserved_realm.enabled", randomBoolean());`

Reply:
Good tip, thanks!
Pushed [2aa73cd](https://github.com/elastic/elasticsearch/pull/106424/commits/2aa73cd563f89025202bcba5a01ecd48303b6a26).